packages/lesswrong/server/twitterBot.ts

```diff
@@ -0,0 +1,134 @@
+import Users from "@/lib/vulcan-users";
+import { dogstatsd } from "./datadog/tracer";
+
+const enabledSetting = new DatabaseServerSetting<boolean>("twitterBot.enabled", false);
+const karmaThresholdSetting = new DatabaseServerSetting<number>("twitterBot.karmaTreshold", 40);
+const apiKeySetting = new DatabaseServerSetting<string | null>("twitterBot.apiKey", null);
+const apiKeySecretSetting = new DatabaseServerSetting<string | null>("twitterBot.apiKeySecret", null);
+const accessTokenSetting = new DatabaseServerSetting<string | null>("twitterBot.accessToken", null);
+const accessTokenSecretSetting = new DatabaseServerSetting<string | null>("twitterBot.accessTokenSecret", null);
+
+const TWEET_MAX_LENGTH = 279;
+const URL_LENGTH = 24;
+
+async function writeTweet(post: DbPost): Promise<string> {
+  const userIds = [
+    post.userId,
+    ...(post.coauthorStatuses ?? []).map(({ userId }) => userId)
+  ];
+
+  const users = await Users.find(
+    { _id: { $in: userIds }, deleted: false }
+  ).fetch();
+
+  const usersOrdered = userIds.map(id => 
+    users.find(user => user._id === id)
+  ).filter(user => user) as DbUser[];
+  const userDisplayNames = usersOrdered.map(user => user.displayName);
+
+  let authorString;
+  switch (userDisplayNames.length) {
+    case 1:
+      authorString = userDisplayNames[0];
+      break;
+    case 2:
+      authorString = `${userDisplayNames[0]} and ${userDisplayNames[1]}`;
+      break;
+    default:
+      authorString = `${userDisplayNames[0]} + ${userDisplayNames.length - 1} others`;
+      break;
+  }
+
+  const preUrlPart = `New popular post from the EA Forum:\n\n"${post.title}" by ${authorString}`;
+  const truncatedPreUrlPart = preUrlPart.length + URL_LENGTH > TWEET_MAX_LENGTH
+    ? preUrlPart.substring(0, TWEET_MAX_LENGTH - URL_LENGTH - 3) + "..."
+    : preUrlPart;
+
+  return `${truncatedPreUrlPart}\n${postGetPageUrl(post, true)}`;
+}
+
+async function postTweet(content: string) {
+  const apiKey = apiKeySetting.get()
+  const apiKeySecret = apiKeySecretSetting.get()
+  const accessToken = accessTokenSetting.get()
+  const accessTokenSecret = accessTokenSecretSetting.get()
+
+  try {
+    if (!apiKey || !apiKeySecret || !accessToken || !accessTokenSecret) {
+      throw new Error("Twitter bot credentials not configured")
+    }
+
+    const twitterClient = new TwitterApi({
+      appKey: apiKey,
+      appSecret: apiKeySecret,
+      accessToken: accessToken,
+      accessSecret: accessTokenSecret,
+    });
+    const rwClient = twitterClient.readWrite
+
+    const { data } = await rwClient.v2.tweet(content);
+    dogstatsd?.increment("tweet_created", 1, 1.0, {outcome: 'success'})
+    return data.id
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(`Error posting tweet. Tweet content: ${content}, error:`, error);
+    dogstatsd?.increment("tweet_created", 1, 1.0, {outcome: 'error'})
+  }
+}
+
+addCronJob({
+  name: "runTwitterBot",
+  interval: "every 31 minutes",
```

Comment:
The main reason I'm not making this more frequent is just in case it breaks in a way that results in a lot of duplicates

Reply:
Just so it doesn't line up on the exact same time every time